src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveTestsWithVariousTypes.cs

```diff
@@ -0,0 +1,238 @@
+            ContentType = AssemblyContentType.WindowsRuntime,
+            Flags = AssemblyNameFlags.EnableJITcompileTracking | AssemblyNameFlags.Retargetable,
+        };
+
+        [Fact]
+        public void EmptyAssemblyAndModuleTest()
+        {
+            using (TempFile file = TempFile.Create())
+            {
+                Assembly assemblyFromDisk = WriteAndLoadAssembly(Type.EmptyTypes, file.Path);
+                Assert.NotNull(assemblyFromDisk);
+
+                AssemblyName aNameFromDisk = assemblyFromDisk.GetName();
+
+                // Test AssemblyName properties
+                Assert.Equal(s_assemblyName.Name, aNameFromDisk.Name);
+                Assert.Equal(s_assemblyName.Version, aNameFromDisk.Version);
+                Assert.Equal(s_assemblyName.CultureInfo, aNameFromDisk.CultureInfo);
+                Assert.Equal(s_assemblyName.CultureName, aNameFromDisk.CultureName);
+                Assert.Equal(s_assemblyName.ContentType, aNameFromDisk.ContentType);
+                // Runtime assemblies adding AssemblyNameFlags.PublicKey in Assembly.GetName() overloads
+                Assert.Equal(s_assemblyName.Flags | AssemblyNameFlags.PublicKey, aNameFromDisk.Flags);
+
+                Module moduleFromDisk = assemblyFromDisk.Modules.First();
+
+                Assert.NotNull(moduleFromDisk);
+                Assert.Equal(s_assemblyName.Name, moduleFromDisk.ScopeName);
+                Assert.Empty(moduleFromDisk.GetTypes());
+            }
+        }
+
+        private static Assembly WriteAndLoadAssembly(Type[] types, string filePath)
+        {
+            AssemblyTools.WriteAssemblyToDisk(s_assemblyName, types, filePath);
+
+            return AssemblyTools.LoadAssemblyFromPath(filePath);
+        }
+
+        public static IEnumerable<object[]> VariousInterfacesTestData()
+        {
+            yield return new object[] { new Type[] { typeof(INoMethod) } };
+            yield return new object[] { new Type[] { typeof(IMultipleMethod) } };
+            yield return new object[] { new Type[] { typeof(INoMethod), typeof(INoMethod2) } };
+            yield return new object[] { new Type[] { typeof(INoMethod), typeof(IOneMethod) } };
+            yield return new object[] { new Type[] { typeof(IMultipleMethod), typeof(INoMethod2) } };
+            yield return new object[] { new Type[] { typeof(IMultipleMethod), typeof(INoMethod2), typeof(IAccess), typeof(IOneMethod), typeof(INoMethod) } };
+        }
+
+        [Theory]
+        [MemberData(nameof(VariousInterfacesTestData))]
+        public void WriteVariousInterfacesToFileTest(Type[] types)
+        {
+            using (TempFile file = TempFile.Create())
+            {
+                Assembly assemblyFromDisk = WriteAndLoadAssembly(types, file.Path);
+                Assert.NotNull(assemblyFromDisk);
+
+                Assert.NotNull(assemblyFromDisk);
+
+                Module moduleFromDisk = assemblyFromDisk.Modules.First();
```

Comment:
The tests have one copy of the code that writes out the types (WriteAndLoadAssembly).

Can we have one copy of the code that compares what was written with the template, to avoid duplicating an ad-hoc comparison in each test?

Reply:
Not for all, because they are testing different members like interface/structs, method/field.

Reply:
The shared implementation can always go over all field and all methods. It is fine if some test cases have no fields or no methods, the shared implementation should just verify the emitted type has no fields or no methods in that case.